devices/Ds1302/Ds1302.cs

```diff
@@ -0,0 +1,186 @@
+            _clockPin.Write(PinValue.Low);
+        }
+
+        /// <summary>
+        /// Checks if the clock is halted.
+        /// </summary>
+        /// <returns><b>true</b> when clock is halted, else <b>false</b></returns>
+        public bool IsHalted()
+        {
+            PrepareRead(Ds1302Registers.REG_SECONDS);
+            byte seconds = ReadByte();
+            EndTransmission();
+            return (byte)(seconds & enableWrite) > 0;
+        }
+
+        /// <summary>
+        /// Halts the clock.
+        /// </summary>
+        public void Halt()
+        {
+            PrepareWrite(Ds1302Registers.REG_SECONDS);
+            WriteByte(enableWrite);
+            EndTransmission();
+        }
+
+        /// <summary>
+        /// Read time from the device.
+        /// </summary>
+        /// <returns>Time from the device.</returns>
+        protected override DateTime ReadTime()
+        {
+            PrepareRead(Ds1302Registers.REG_BURST);
+            var seconds = NumberHelper.Bcd2Dec((byte)(ReadByte() & 0b0111_1111));
+            var minutes = NumberHelper.Bcd2Dec((byte)(ReadByte() & 0b0111_1111));
+            var hours = NumberHelper.Bcd2Dec((byte)(ReadByte() & 0b0011_1111));
+            var days = NumberHelper.Bcd2Dec((byte)(ReadByte() & 0b0011_1111));
+            var months = NumberHelper.Bcd2Dec((byte)(ReadByte() & 0b0001_1111));
+            var dayOfWeek = NumberHelper.Bcd2Dec((byte)(ReadByte() & 0b0000_0111));
+            var years = NumberHelper.Bcd2Dec((byte)(ReadByte() & 0b0111_1111));
+
+            EndTransmission();
+
+            return new DateTime(2000 + years, months, days, hours, minutes, seconds, 0);
+        }
+
+        /// <summary>
+        /// Sets date and time.
+        /// </summary>
+        /// <param name="time">Date and time.</param>
+        protected override void SetTime(DateTime time)
+        {
+            PrepareWrite(Ds1302Registers.REG_WP);
+            WriteByte(disableWrite);
+            EndTransmission();
+
+            PrepareWrite(Ds1302Registers.REG_BURST);
+            WriteByte(NumberHelper.Dec2Bcd(time.Second));
+            WriteByte(NumberHelper.Dec2Bcd(time.Minute));
+            WriteByte(NumberHelper.Dec2Bcd(time.Hour));
+            WriteByte(NumberHelper.Dec2Bcd(time.Day));
+            WriteByte(NumberHelper.Dec2Bcd(time.Month));
+            WriteByte(NumberHelper.Dec2Bcd((int)time.DayOfWeek));
+            WriteByte(NumberHelper.Dec2Bcd(time.Year - 2000));
+
+            WriteByte(enableWrite);
+            EndTransmission();
+        }
+
+        /// <summary>
+        /// Releases the unmanaged resources used by the RtcBase and optionally releases the managed resources.
+        /// </summary>
+        /// <param name="disposing">True to release both managed and unmanaged resources; false to release only unmanaged resources.</param>
+        protected override void Dispose(bool disposing)
+        {
+            _dataPin?.Dispose();
```

Comment:
see the pattern as well with the _shouldDispose. In all cases, you'll have to dispose the GpioPin, and the controller if asked for it. With your current pattern, it's disposed all the time and if it is pass to the binding and used for something else, it won't work properly.